File: vbox-app/src/main/java/com/kedzie/vbox/machine/settings/StorageFragment.java
package com.kedzie.vbox.machine.settings;

import android.content.res.Configuration;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;

import com.kedzie.vbox.R;
import com.kedzie.vbox.api.IMedium;
import com.kedzie.vbox.api.IStorageController;
import com.kedzie.vbox.api.jaxb.DeviceType;
import com.kedzie.vbox.api.jaxb.IMediumAttachment;
import com.kedzie.vbox.app.BundleBuilder;
import com.kedzie.vbox.app.FragmentElement;
import com.kedzie.vbox.app.Utils;
import com.kedzie.vbox.machine.settings.StorageListFragment.OnMediumAttachmentClickedListener;
import com.kedzie.vbox.machine.settings.StorageListFragment.OnStorageControllerClickedListener;


public class StorageFragment extends Fragment implements OnStorageControllerClickedListener, OnMediumAttachmentClickedListener {
    private static final String TAG = "StorageFragment";
    private boolean _dualPane;
    private StorageListFragment _listFragment;
    
    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
    }
    
	@Override
	public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        FrameLayout view = new FrameLayout(getActivity());
	    LayoutInflater.from(getActivity()).inflate(R.layout.settings_storage, view, true);
        FrameLayout detailsFrame = (FrameLayout)view.findViewById(R.id.details);
        _dualPane = detailsFrame != null && detailsFrame.getVisibility() == View.VISIBLE;
        
        Fragment f = getChildFragmentManager().findFragmentByTag("list");
        if(f==null) {
            _listFragment = new StorageListFragment();
            _listFragment.setArguments(getArguments());
            getChildFragmentManager().beginTransaction().add(R.id.list, _listFragment, "list").commit();
        } else {
            _listFragment = (StorageListFragment)f;
        }
        return view;
	}

    @Override
    public void onConfigurationChanged(Configuration newConfig) {
        super.onConfigurationChanged(newConfig);
        if(Utils.getScreenSize(newConfig)==Configuration.SCREENLAYOUT_SIZE_XLARGE) {
            Log.i(TAG, "Handling orientation change");
            FragmentManager mgr = getChildFragmentManager();
            FragmentTransaction tx = mgr.beginTransaction();
            for(Fragment fragment : mgr.getFragments()) {
                tx.detach(fragment);
            }
            tx.commit();
            FrameLayout view = (FrameLayout)getView();
            view.removeAllViews();
            LayoutInflater.from(getActivity()).inflate(R.layout.settings_storage, view, true);
            FrameLayout detailsFrame = (FrameLayout)view.findViewById(R.id.details);
            _dualPane = detailsFrame != null && detailsFrame.getVisibility() == View.VISIBLE;

            _listFragment = (StorageListFragment) getChildFragmentManager().findFragmentByTag("list");
            Fragment detailsFragment = getChildFragmentManager().findFragmentByTag("details");
            tx = getChildFragmentManager().beginTransaction();
            tx.attach(_listFragment);
            if(detailsFragment!=null && _dualPane) {
                tx.remove(detailsFragment).add(R.id.details, Fragment.instantiate(getActivity(), detailsFragment.getClass().getName(), detailsFragment.getArguments()), "details");
            }
            tx.commit();
        }
    }

    @Override
    public void onStorageControllerClicked(IStorageController element) {
        show(new FragmentElement(element.getName(), R.drawable.ic_settings_storage, StorageControllerFragment.class, new BundleBuilder().putAll(getArguments()).putParcelable(IStorageController.BUNDLE, element).create()));
    }
    
    @Override
	public void onMediumAttachmentClicked(IMediumAttachment element) {
    	FragmentElement details = new FragmentElement("Attachment", R.drawable.ic_settings_storage, null, new BundleBuilder().putAll(getArguments()).putParcelable(IMedium.BUNDLE, element).create());
    	if(element.getType().equals(DeviceType.HARD_DISK))
    		details.clazz = StorageHardDiskFragment.class;
    	else if(element.getType().equals(DeviceType.DVD))
    		details.clazz = StorageDVDFragment.class;
//    	else if(element.getType().equals(DeviceType.FLOPPY))
//    		details.clazz = StorageFloppyFragment.class;
        show(details);
	}
    
    private void show(FragmentElement details) {
    	if(_dualPane) {
    		Utils.setCustomAnimations(getChildFragmentManager().beginTransaction()).replace(R.id.details, details.instantiate(getActivity())).commit();
        } else {
            Utils.setCustomAnimations(getChildFragmentManager().beginTransaction())
                    .addToBackStack(null)
                    .detach(_listFragment)
                    .add(R.id.list, details.instantiate(getActivity()), "details")
                    .commit();
        }
    }


    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        if(_listFragment !=null)
            _listFragment.onCreateOptionsMenu(menu, inflater);
    }
    
    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if(_listFragment != null)
             return _listFragment.onOptionsItemSelected(item);
        return false;
    }
}

File: vbox-app/src/main/java/com/kedzie/vbox/machine/settings/StorageFragment.kt
package com.kedzie.vbox.machine.settings

import android.content.res.Configuration
import android.os.Bundle
import android.view.*
import android.widget.FrameLayout
import androidx.fragment.app.Fragment
import com.kedzie.vbox.R
import com.kedzie.vbox.api.IMedium
import com.kedzie.vbox.api.IStorageController
import com.kedzie.vbox.api.jaxb.DeviceType
import com.kedzie.vbox.api.jaxb.IMediumAttachment
import com.kedzie.vbox.app.BundleBuilder
import com.kedzie.vbox.app.FragmentElement
import com.kedzie.vbox.app.Utils
import com.kedzie.vbox.machine.settings.StorageListFragment.OnMediumAttachmentClickedListener
import com.kedzie.vbox.machine.settings.StorageListFragment.OnStorageControllerClickedListener
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import timber.log.Timber


class StorageFragment : Fragment(), CoroutineScope by CoroutineScope(Dispatchers.Main),
        OnStorageControllerClickedListener, OnMediumAttachmentClickedListener {
    private var _dualPane: Boolean = false
    private var _listFragment: StorageListFragment? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = FrameLayout(activity!!)
        LayoutInflater.from(activity).inflate(R.layout.settings_storage, view, true)
        val detailsFrame = view.findViewById<View>(R.id.details) as FrameLayout
        _dualPane = detailsFrame != null && detailsFrame.visibility == View.VISIBLE

        val f = childFragmentManager.findFragmentByTag("list")
        if (f == null) {
            _listFragment = StorageListFragment()
            _listFragment!!.arguments = arguments
            childFragmentManager.beginTransaction().add(R.id.list, _listFragment!!, "list").commit()
        } else {
            _listFragment = f as StorageListFragment?
        }
        return view
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        if (Utils.getScreenSize(newConfig) == Configuration.SCREENLAYOUT_SIZE_XLARGE) {
            Timber.d( "Handling orientation change")
            val mgr = childFragmentManager
            var tx = mgr.beginTransaction()
            for (fragment in mgr.fragments) {
                tx.detach(fragment)
            }
            tx.commit()
            val view = view as FrameLayout?
            view!!.removeAllViews()
            LayoutInflater.from(activity).inflate(R.layout.settings_storage, view, true)
            val detailsFrame = view.findViewById<View>(R.id.details) as FrameLayout
            _dualPane = detailsFrame != null && detailsFrame.visibility == View.VISIBLE

            _listFragment = childFragmentManager.findFragmentByTag("list") as StorageListFragment?
            val detailsFragment = childFragmentManager.findFragmentByTag("details")
            tx = childFragmentManager.beginTransaction()
            tx.attach(_listFragment!!)
            if (detailsFragment != null && _dualPane) {
                tx.remove(detailsFragment).add(R.id.details, Fragment.instantiate(activity!!, detailsFragment.javaClass.name, detailsFragment.arguments), "details")
            }
            tx.commit()
        }
    }

    override fun onStorageControllerClicked(element: IStorageController) {
        launch {
            show(FragmentElement(element.getName(), R.drawable.ic_settings_storage, StorageControllerFragment::class.java, BundleBuilder().putAll(arguments).putParcelable(IStorageController.BUNDLE, element).create()))
        }
    }

    override fun onMediumAttachmentClicked(element: IMediumAttachment) {
        val details = FragmentElement("Attachment",
                R.drawable.ic_settings_storage,
                if (element.deviceType == DeviceType.HARD_DISK)
                    StorageHardDiskFragment::class.java
                else StorageDVDFragment::class.java,
                Bundle().apply {
                    putAll(arguments)
                    putParcelable(IMedium.BUNDLE, element)
                })

        //    	else if(element.getType().equals(DeviceType.FLOPPY))
        //    		details.clazz = StorageFloppyFragment.class;
        show(details)
    }

    private fun show(details: FragmentElement) {
        if (_dualPane) {
            Utils.setCustomAnimations(childFragmentManager.beginTransaction()).replace(R.id.details, details.instantiate(activity)).commit()
        } else {
            Utils.setCustomAnimations(childFragmentManager.beginTransaction())
                    .addToBackStack(null)
                    .detach(_listFragment!!)
                    .add(R.id.list, details.instantiate(activity), "details")
                    .commit()
        }
    }


    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        if (_listFragment != null)
            _listFragment!!.onCreateOptionsMenu(menu, inflater)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return if (_listFragment != null) _listFragment!!.onOptionsItemSelected(item) else false
    }

    override fun onDestroy() {
        cancel()
        super.onDestroy()
    }
}
